File: cs/doctor.py
"""`cs doctor`: environment and consistency checks with fix hints."""
from __future__ import annotations

import json
import shutil
import sys
from pathlib import Path
from typing import Callable, List, Tuple

from . import apply, gitutil, link, paths, platform
from .config import Machine
from .manifest import Manifest
from .ui import fail, info, ok, warn

Result = Tuple[str, str]   # ("ok"|"warn"|"fail", message)


def _check_python() -> Result:
    v = sys.version_info
    return ("ok", f"python {v.major}.{v.minor}") if v >= (3, 9) else ("fail", "python < 3.9")


def _check_git() -> Result:
    v = gitutil.version()
    if v < (2, 36):
        return ("warn", f"git {'.'.join(map(str, v))} < 2.36: includeIf hasconfig unsupported; identities fall back to per-repo config")
    return ("ok", f"git {'.'.join(map(str, v))}")


def _check_claude() -> Result:
    return ("ok", f"claude at {shutil.which('claude')}") if shutil.which("claude") else ("warn", "claude not on PATH (curl -fsSL https://claude.ai/install.sh | bash)")


def _check_platform() -> Result:
    platform.refuse_unsupported()
    return ("ok", platform.describe())


def _check_mnt(man: Manifest, m: Machine) -> Result:
    ws = man.workspace(m)
    if platform.is_wsl() and str(ws).startswith("/mnt/"):
        return ("fail", f"workspace {ws} is on the Windows filesystem; use the WSL ext4 home")
    return ("ok", f"workspace {paths.contract(ws)}")


def _check_links(repo: Path) -> Result:
    cdir = paths.claude_dir()
    broken = [i for i in apply.LINK_ITEMS + ["plans"] if (cdir / i).is_symlink() and not (cdir / i).exists()]
    if broken:
        return ("fail", "broken links in ~/.claude: " + ", ".join(broken) + "  (cs apply)")
    return ("ok", "~/.claude links healthy")


def _check_settings(repo: Path, m: Machine, man: Manifest) -> Result:
    changes: List[str] = []
    apply.apply_settings(repo, m, True, changes)
    return ("warn", "settings.json drift: " + "; ".join(changes) + "  (cs apply)") if changes else ("ok", "settings.json rendered")


def _check_identities(repo: Path, m: Machine, man: Manifest) -> List[Result]:
    out: List[Result] = []
    ws = man.workspace(m)
    for p in man.selected(m):
        root = p.checkout_root(ws)
        if p.kind != "git" or not root.exists() or not gitutil.is_repo(root):
            continue
        ident = man.identities.get(p.identity)
        email = gitutil.config_get(root, "user.email")
        url = gitutil.remote_url(root)
        if p.url and gitutil.canonical_github(url) != gitutil.canonical_github(p.url):
            out.append(("warn", f"{p.name}: remote {url} ≠ manifest {p.url}  (cs doctor --fix)"))
        elif url and url != p.url and p.url:
            out.append(("warn", f"{p.name}: remote uses alias/other form {url}; manifest {p.url}  (cs doctor --fix)"))
        if ident and email != ident.email:
            out.append(("fail", f"{p.name}: user.email resolves to '{email or 'UNSET'}', expected {ident.email}"))
    return out or [("ok", "git identities resolve per manifest")]


def _check_local_mcp_secrets(man: Manifest, m: Machine) -> Result:
    cj = paths.claude_json()
    if not cj.exists():
        return ("ok", "no ~/.claude.json")
    try:
        data = json.loads(cj.read_text())
    except ValueError:
        return ("warn", "~/.claude.json unparsable")
    hits = []
    for path, entry in data.get("projects", {}).items():
        for name, cfg in (entry.get("mcpServers") or {}).items():
            if cfg.get("env") or cfg.get("headers"):
                hits.append(f"{name}@{paths.contract(Path(path))}")
    if hits:
        return ("warn", "local-scope MCP servers with env/headers in ~/.claude.json (not portable): " + ", ".join(hits) + "  (cs adopt mcp <project>)")
    return ("ok", "no secret-bearing local-scope MCP servers")


def _check_gh_token_env() -> Result:
    import os
    if os.environ.get("GH_TOKEN") or os.environ.get("GITHUB_TOKEN"):
        return ("warn", "GH_TOKEN/GITHUB_TOKEN is exported in this shell; gh ignores its stored logins while set")
    return ("ok", "no GH_TOKEN override in env")


def fix(repo: Path, m: Machine, man: Manifest) -> None:
    ws = man.workspace(m)
    for p in man.selected(m):
        root = p.checkout_root(ws)
        if p.kind != "git" or not root.exists() or not gitutil.is_repo(root) or not p.url:
            continue
        url = gitutil.remote_url(root)
        if url != p.url and gitutil.canonical_github(url) == gitutil.canonical_github(p.url):
            gitutil.run(["remote", "set-url", "origin", p.url], root)
            ok(f"{p.name}: remote url -> {p.url}")
    apply.run(repo, m, man)


def run(repo: Path, m: Machine, man: Manifest, do_fix: bool = False) -> int:
    if do_fix:
        fix(repo, m, man)
    results: List[Result] = [
        _check_platform(), _check_python(), _check_git(), _check_claude(),
        _check_mnt(man, m), _check_links(repo), _check_settings(repo, m, man),
        _check_local_mcp_secrets(man, m), _check_gh_token_env(),
    ] + _check_identities(repo, m, man)
    rc = 0
    for level, msg in results:
        {"ok": ok, "warn": warn, "fail": fail}[level](msg)
        if level == "fail":
            rc = 1
    return rc

File: src/doctor.ts
/** cs doctor [--fix]: environment and consistency checks. */
import { existsSync, lstatSync, readdirSync, readFileSync } from "node:fs";
import { join } from "node:path";
import * as git from "./git.js";
import { claudeDir, claudeJson, contract } from "./paths.js";
import * as platform from "./platform.js";
import type { Machine } from "./machine.js";
import { identityMatches, loadManifest, selectedProjects, workspace, type Manifest, type Project } from "./manifest.js";
import { locate, present } from "./checkout.js";
import { applySettings } from "./apply.js";
import * as ui from "./ui.js";
import { which } from "./deps.js";
import { unregisteredDirs } from "./status.js";
import { ago } from "./plan.js";
import { HOOK_EVENTS, hooksStatus } from "./hooks.js";
import { add, fixRemote } from "./projects.js";
import { findOldNames, migrate } from "./migrate.js";
import { orphans } from "./remove.js";

type R = ["ok" | "warn" | "fail", string];
const LINKS = ["CLAUDE.md", "rules", "agents", "themes", "keybindings.json", "plans"];

/** Registered projects (present here) and workspace directories that have no remote yet — the ensure-remote candidates. */
export function remoteless(man: Manifest, m: Machine): { projects: Project[]; dirs: { name: string; remote: string }[] } {
  const ws = workspace(man, m);
  const projects = selectedProjects(man, m).filter((p) => { const c = locate(p, ws); return present(c) ? !p.url : c.why !== "missing"; });
  return { projects, dirs: unregisteredDirs(ws, new Set(Object.values(man.projects).map((p) => p.path || p.name))) };
}
/** Returns the share's path: the old-name moves may have relocated it. */
export async function fix(repo: string, m: Machine, man: Manifest): Promise<string> {
  const mig = migrate(repo, m, man); for (const d of mig.done) ui.ok(d);
  if (mig.repo !== repo) { repo = mig.repo; man = loadManifest(repo); }
  const ws = workspace(man, m);
  const { projects, dirs } = remoteless(man, m);
  for (const p of projects) {
    if (!ui.canAsk()) { ui.warn(`${p.name}: no remote — run cs doctor --fix in a terminal to create one`); continue; }
    if (await ui.confirm(`${p.name} has no remote — create a private GitHub repo and push it?`, true)) { if (await fixRemote(repo, m, man, p)) man = loadManifest(repo); }
  }
  for (const d of dirs) {
    const path = join(ws, d.name);
    if (!ui.canAsk()) { ui.warn(`${d.name}: not registered — cs add ${contract(path)}`); continue; }
    if (await ui.confirm(`${d.name} is not registered — register it${d.remote ? "" : " (creating a private GitHub repo)"}?`, true)) { try { await add(repo, m, man, path, { profiles: [], description: "", noCommit: false }); man = loadManifest(repo); } catch (e: any) { ui.fail(e.message); } }
  }
  for (const p of selectedProjects(man, m)) {
    const c = locate(p, ws); if (!present(c) || !p.url) continue; const root = c.root;
    const url = git.remoteUrl(root); if (url === p.url) continue;
    const cur = git.canonicalGithub(url), want = git.canonicalGithub(p.url);
    let same = cur === want; const ident = p.identity ? man.identities[p.identity] : undefined;
    if (!same && ident && identityMatches(ident, cur) && cur.split("/").pop() === want.split("/").pop()) same = true;
    if (same) { git.git(["remote", "set-url", "origin", p.url], root); ui.ok(`${p.name}: remote url ${url} → ${p.url}`); }
    else ui.warn(`${p.name}: remote ${url} is a different repo than manifest ${p.url}; not changing it`);
  }
  return repo;
}
export async function runDoctor(repo: string, m: Machine, man: Manifest, doFix = false, compact = false): Promise<number> {
  if (doFix) { repo = await fix(repo, m, man); man = loadManifest(repo); }
  const res: R[] = [];
  platform.refuseUnsupported(); res.push(["ok", platform.describe()]);
  res.push(["ok", `node ${process.versions.node}`]);
  const v = git.version(); res.push(v[0] > 2 || (v[0] === 2 && v[1] >= 36) ? ["ok", `git ${v.join(".")}`] : ["warn", `git ${v.join(".")} < 2.36: identities fall back to per-repo config`]);
  res.push(which("claude") ? ["ok", `claude at ${which("claude")}`] : ["warn", "claude not on PATH (curl -fsSL https://claude.ai/install.sh | bash)"]);
  const ws = workspace(man, m); res.push(platform.isWSL() && ws.startsWith("/mnt/") ? ["fail", `workspace ${ws} is on the Windows filesystem; use the WSL home`] : ["ok", `workspace ${contract(ws)}`]);
  const broken = LINKS.filter((i) => { try { return lstatSync(join(claudeDir(), i)).isSymbolicLink() && !existsSync(join(claudeDir(), i)); } catch { return false; } });
  res.push(broken.length ? ["fail", "broken links in ~/.claude: " + broken.join(", ") + "  (cs apply)"] : ["ok", "~/.claude links healthy"]);
  const ch: string[] = []; applySettings(repo, m, true, ch); res.push(ch.length ? ["warn", "settings.json drift: " + ch.join("; ") + "  (cs apply)"] : ["ok", "settings.json rendered"]);
  if (existsSync(claudeJson())) { try { const d = JSON.parse(readFileSync(claudeJson(), "utf8")); const hits: string[] = [];
    for (const [path, e] of Object.entries<any>(d.projects ?? {})) for (const [n, c] of Object.entries<any>(e.mcpServers ?? {})) if (c.env || c.headers) hits.push(`${n}@${contract(path)}`);
    res.push(hits.length ? ["warn", `local-scope MCP servers with secrets in ~/.claude.json (machine-only): ${hits.join(", ")} — keep until cs secrets provides the \${VAR}s, then \`claude mcp remove <name> -s local\``] : ["ok", "no secret-bearing local-scope MCP servers"]); } catch { res.push(["warn", "~/.claude.json unparsable"]); } }
  res.push(process.env.GH_TOKEN || process.env.GITHUB_TOKEN ? ["warn", "GH_TOKEN/GITHUB_TOKEN is exported in this shell; gh ignores its stored logins while set"] : ["ok", "no GH_TOKEN override in env"]);
  const idr: R[] = [];
  for (const p of selectedProjects(man, m)) { const c = locate(p, ws); if (!present(c) && c.why !== "no remote") continue; const root = c.root;   // a remote-less repo still has an identity to check
    const ident = p.identity ? man.identities[p.identity] : undefined; const email = git.configGet(root, "user.email"); const url = git.remoteUrl(root);
    if (p.url && git.canonicalGithub(url) !== git.canonicalGithub(p.url)) idr.push(["warn", `${p.name}: remote ${url} ≠ manifest ${p.url}  (cs doctor --fix)`]);
    else if (url && p.url && url !== p.url) idr.push(["warn", `${p.name}: remote uses alias/other form ${url}; manifest ${p.url}  (cs doctor --fix)`]);
    if (ident && email !== ident.email) idr.push(["fail", `${p.name}: user.email resolves to '${email || "UNSET"}', expected ${ident.email}`]); }
  res.push(...(idr.length ? idr : [["ok", "git identities resolve per manifest"] as R]));
  const hs = hooksStatus(repo);
  res.push(hs.complete ? ["ok", `hooks installed (${HOOK_EVENTS.join(", ")})`] : ["warn", `hooks ${hs.events.length ? "outdated" : "not installed"}  (cs sync re-installs them)`]);
  res.push(hs.timerActive ? ["ok", "timer active (share sync every 15 min)"] : hs.timerFiles ? ["warn", "timer installed but not active  (cs sync re-installs it)"] : ["warn", "timer not installed  (cs sync installs it)"]);
  res.push(hs.lastSync ? ["ok", `last share sync ${ago(hs.lastSync)}`] : ["warn", "the share has never synced here  (cs sync)"]);
  if (m.secretsBackend !== "none" && existsSync(join(repo, ".sops.yaml"))) {
    const machines = existsSync(join(repo, "machines")) ? readdirSync(join(repo, "machines")).filter((d) => existsSync(join(repo, "machines", d, "age.pub"))) : [];
    if (!machines.includes("recovery")) res.push(["warn", "secrets have no recovery key — cs secrets recovery (print it once, keep it in your password manager)"]);
    const others = machines.filter((d) => d !== m.name && d !== "recovery");
    if (others.length) res.push(["ok", `machines trusted with the secrets: ${others.join(", ")}  (cs untrust <machine> when one is retired)`]);
  }
  const old = findOldNames(repo, m, man);
  for (const o of old) res.push(["warn", `old name — ${o.what}: ${o.move}${o.then ? `; ${o.then}` : ""}  ${o.apply ? "(cs doctor --fix; docs/MIGRATION.md)" : "(by hand; docs/MIGRATION.md)"}`]);
  if (!old.length) res.push(["ok", "on-disk names current (share key, share, handoffs, state)"]);
  const rl = remoteless(man, m);
  for (const p of rl.projects) res.push(["fail", `${p.name}: no remote  (cs doctor --fix · or cs remove ${p.name})`]);
  for (const d of rl.dirs) res.push(["warn", `${contract(join(ws, d.name))}: not registered${d.remote ? "" : ", no remote"}  (cs add ${contract(join(ws, d.name))})`]);
  if (!rl.projects.length && !rl.dirs.length) res.push(["ok", "every project has a remote; nothing unregistered under the workspace"]);
  const left = orphans(repo, man);   // state or secrets for a name no longer in the manifest: never fixed by --fix, removal always asks (or takes --yes)
  for (const n of left) res.push(["warn", `${n}: state/secrets in the share but not registered  (cs remove ${n})`]);
  if (!left.length) res.push(["ok", "the share holds state and secrets for registered projects only"]);
  const sym = { ok: ui.green("✓"), warn: ui.yellow("!"), fail: ui.red("✗") };
  if (compact) { const bad = res.filter(([l]) => l !== "ok"); if (bad.length) ui.table(bad.map(([l, msg]) => [sym[l], msg])); ui.step(`${res.length - bad.length} of ${res.length} checks passed`); }
  else ui.table(res.map(([l, msg]) => [sym[l], msg]));
  return res.some(([l]) => l === "fail") ? 1 : 0;
}
